OrbitQt/TutorialOverlay.h

```diff
@@ -0,0 +1,166 @@
+  dimming overlay and cannot be interacted with
+* An init callback: A callback method that is executed when the step is started. This can
+  be used to e.g. connect signals to the TutorialOverlay::NextStep() method to
+  automatically advance
+* A teardown callback: A callback method that is executed when the step ends. Use this to
+  cleanup everything you set up in the init callback
+* A verify callback: A callback method that is executed before a step is finished. If this
+  callback returns a string, this is interpreted as an error message to be displayed to
+  the user, and the tutorial will not advance to the next step.
+
+Check the unit tests on how to use.
+A very brief summary:
+1) Add your step's UI to tutorialoverlay.ui
+2) (optional) Setup your step with SetupStep() in code
+3) Create a section that contains one or more steps with AddSection()
+4) Create this dialog with the main window as a parent
+5) Execute the tutorial by StartSection()
+**/
+class TutorialOverlay : public QDialog {
+  Q_OBJECT
+
+ public:
+  using StepCallback = std::function<void(TutorialOverlay* overlay, const std::string& step_name)>;
+  using VerifyStepCompleted = std::function<std::optional<std::string>(
+      TutorialOverlay* overlay, const std::string& step_completed)>;
+
+  struct StepSetup {
+    QWidget* anchor_widget = nullptr;
+    StepCallback callback_init = nullptr;
+    StepCallback callback_teardown = nullptr;
+    VerifyStepCompleted callback_verify = nullptr;
+  };
+
+  explicit TutorialOverlay(QWidget* parent);
+  ~TutorialOverlay() override;
+
+  bool eventFilter(QObject* object, QEvent* event) override;
+
+  void SetupStep(const std::string& ui_tab_name, StepSetup step_setup);
+  [[nodiscard]] const StepSetup& GetStep(const std::string name) const;
+
+  [[nodiscard]] std::string GetActiveStepName() const {
+    return HasActiveStep() ? GetActiveStep().name : "";
+  }
+
+  std::string GetActiveSectionName() const {
+    return HasActiveSection() ? active_section_->first : "";
+  }
+
+  void AddSection(std::string section_name, std::string title, std::vector<std::string> step_names);
+
+ public Q_SLOTS:
+  void StartSection(const std::string& section_name);
+  void NextStep();
+  void PrevStep();
+
+ Q_SIGNALS:
+  void SectionCompleted(const std::string section_name);
+
+ protected:
+  void closeEvent(QCloseEvent* event) override;
+  void showEvent(QShowEvent* event) override;
+
+ private:
+  enum class HintAnchor { kTopLeft, kTopRight, kBottomRight, kBottomLeft };
+  struct Hint {
+    QWidget* widget = nullptr;
+    HintAnchor anchor = HintAnchor::kTopLeft;
+    QPoint offset = QPoint(0, 0);
+  };
+
+  struct Step {
+    StepSetup setup;
+    std::vector<Hint> hints;
+    QWidget* cutout_widget = nullptr;
+    int tab_index = -1;
+    std::string name;
+  };
+
+  struct Section {
+    std::string title;
+    std::vector<std::string> step_names;
+    std::vector<std::string>::iterator active_step = step_names.end();
+  };
```

Comment:
Storing the iterator seems dangerous to me. The iterator will become invalid on vector reallocation. Could active_step be an index instead? Maybe we should encapsulate those fields also and provide "IncrementActiveStep" and "DecrementActiveStep" methods?

Reply:
Right, I forgot that the iterator will become invalid in this case. It can't really happen because the API only allows you to add a complete section at once, so `step_names` will never be modified. Nevertheless I can see this becoming a problem if we decide to add some functionality on section management later, so I changed that.

I didn't add the Increment/Decrement methods though, because this is kind of what the PrevStep / NextStep is for - granted, they do a bit more, but otherwise I'd need to duplicate some logic or prepare the Increment / Decrement in a way that they're only used by Prev/NextStep anyways...